qubes/ext/audio.py

```diff
@@ -0,0 +1,93 @@
+#
+# The Qubes OS Project, https://www.qubes-os.org/
+#
+# Copyright (C) 2019 Frédéric Pierret <frederic.pierret@qubes-os.org>
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, see <https://www.gnu.org/licenses/>.
+#
+
+import asyncio
+
+import qubes.config
+import qubes.ext
+
+
+class AUDIO(qubes.ext.Extension):
+    # pylint: disable=unused-argument,no-self-use
+    @staticmethod
+    def attached_vms(vm):
+        for domain in vm.app.domains:
+            if getattr(domain, 'audiovm', None) and domain.audiovm == vm:
+                yield domain
+
+    @qubes.ext.handler('domain-pre-shutdown')
+    @asyncio.coroutine
+    def on_domain_pre_shutdown(self, vm, event, **kwargs):
+        attached_vms = [domain for domain in self.attached_vms(vm) if
+                        domain.is_running()]
+        if attached_vms and not kwargs.get('force', False):
+            raise qubes.exc.QubesVMError(
+                self, 'There are running VMs using this VM as AudioVM: '
+                      '{}'.format(', '.join(vm.name for vm in attached_vms)))
+
+    @qubes.ext.handler('domain-init', 'domain-load')
+    def on_domain_init_load(self, vm, event):
+        if getattr(vm, 'audiovm', None):
+            if 'audiovm-' + vm.audiovm.name not in list(vm.tags):
```

Comment:
You don't need to convert tags into a list, it should be enough to use `not in vm.tags`.
It is necessary in few other places, if you add/remove them while iterating.